### PR TITLE
Add: The Runestone target.

### PR DESCRIPTION
**Pull request by bjones1**
Add a Runestone format to the project file, making it much easier to build for Runestone.

### CODE REVIEW

**Comment by bjones1** [comment]
Note that tests fail, but not (necessarily) because of this PR -- we still have lots of other fixing to do.

File: pretext/project/__init__.py [hunk 1]
```diff
@@ -85,61 +86,49 @@ class Target(pxml.BaseXmlModel, tag="target", search_mode="unordered"):
...
+        # When the format is Runestone, this is overwritten in `post_validate`. Make sure it's not specified.
+        if values["format"] == Format.RUNESTONE and v is not None:
+            raise ValueError("The Runestone format's output-dir must not be specified.")
         return Path(v) if v is not None else Path(values["name"])
 
-    # We validate compression before output_filename to use its value to check if we can have an output_filename
+    # We validate compression before output_filename to use its value to check if we can have an output_filename.
     compression: t.Optional[Compression] = pxml.attr()
 
+    # Compression is only supported for HTML.
```
Comment (oscarlevin) [inline pretext/project/__init__.py:103]: And webwork (i.e., webwork-sets), the build target, not the asset

**Review by oscarlevin — CHANGES_REQUESTED** [review]
Looks great.

File: pretext/project/__init__.py (hunk 1 above)
Comment (bjones1) [reply pretext/project/__init__.py:103]: Is this correct?

``` Python
    # Compression is only supported for HTML and WeBWorK formats.
    @validator("compression")
    def compression_validator(
        cls, v: t.Optional[Compression], values: t.Any
    ) -> t.Optional[Compression]:
        if values["format"] not in (Format.HTML, Format.WEBWORK) and v is not None:
            raise ValueError("Only the HTML and WeBWorK formats support compression.")
        return v
```

**Comment by StevenClontz** [comment]
I missed the discussion yesterday, but something to consider if it hasn't already: if the object is for an author to only worry about a working HTML-format target (either the default "web" name, or a custom "runestone" name), and the only time a Runestone version is built is when Brad's book server builds it, would it make sense for us to concentrate on just providing the hooks necessary to modify a working HTML-format target, and let the implementation happen on the book server side?

E.g. the book server gets the book from Github, grabs (via the Project() api) the "runestone" named target if it exists, and uses "web" if not. Then the format can be coerced to HTML (or error if not HTML), a copy of the publication.ptx file can be made, modified to ensure the runestone method is used for html, and the target's publication path can be rewritten to point to this copy. Finally the output-dir is pointed to wherever it needs to be pointed.

**Comment by bnmnetp** [comment]
@StevenClontz you can see a summary here #513 

My favorite option is described there.  An author works with regular html and then deploys to runestone at that point. we either add a Runestone target or make one up programmatically and then copy the publisher file, or just use the publisher file for html and change the platform programmatically .

I think we want to preserve the "old" option of having an explicit runestone target and publisher file for people like Sean who choose to run their own server.  But don't want or need to mess with the author server infrastructure.

I will need to create some kind of api, that accepts the request to deploy, and validates some kind of token (I still want to grant permission to publish on runestone, not just let anyone deploy anything there). On my side I would then handle the details of cloning to the right place, making sure the document-id is valid and available, adding an entry to the library table etc.

**Comment by StevenClontz** [comment]
One other possibility that doesn't require modifying an author's publication file: what would be the result of having a single Runestone publication file that's used for all books hosted on Runestone? If Runestone is the publisher, why are authors allowed to make publication decisions for that platform? (I expect it's not so simple in practice, but I think this might fit with the underlying theory for having a separate publication file in the first place.)

**Comment by bnmnetp** [comment]
@StevenClontz we did talk about that, but there are many options wrt knowls and other things that authors do set.  Oscar has custom CSS and custom XSL (as do runestone RST authors) so I don't think the "one publication file to rule them all" approach is feasible.

**Comment by bnmnetp** [comment]
Having just said that, I would say that custom xsl feels like an author option not a publisher option.  I want to allow custom xsl especially for author/researchers who want to add new interactive elements that may not be relevant to any text but their own and so do not belong in pretext.

File: pretext/project/__init__.py (hunk 1 above)
Comment (oscarlevin) [reply pretext/project/__init__.py]: Looks good!

**Comment by oscarlevin** [comment]
Custom xsl could be used to modify the look/feel of a book, making it very much a publisher option.  But it can also be used to support additional features the author has implemented.  So both authors and publishers would be interested in custom xsl.

**Review by oscarlevin — APPROVED** [review]
This looks great.  After our discussion yesterday, I think we need to do more, but let's merge this in so we have everything in once place.

**Comment by StevenClontz** [comment]
Just ignore me if I'm beating a horse y'all already killed. :-) But custom XSL is controlled by project.ptx, so that's not precluded by providing an offiical Runestone publication file. 

CSS was maybe what I had in mind earlier when suggesting practice vs theory: if CSS is an option an author should get to choose, why is it in the publication file? If there's a very small subset of publisher choices an author is allowed to make when hosting on Runestone, we could let them set them via either metadata in project.ptx or via the Runestone book server ("choose your publication file" from a finite list, or create a publication file dynamically from options chosen).

I do think that having Runestone be more opinionated about publication settings is a major change from the status quo that can wait though.

**Pull request merged**